### ISSUE TITLE
FieldArray remove and push immediately cause view not rerender

### ISSUE THREAD

**Issue by sckoh** [erikras/redux-form#2466]
I have a method that remove a field from FieldArray and immediately add a field into FieldArray but the view is not rerender. 
The state in redux is already updated to the new field when I look at the redux logger log.

```
  fields.remove(0);
  fields.push({ postId: 1 });
```

When I press on one of the input in the FieldArray, the FieldArray will only rerender.

After I read through the source code of ConnectedFieldArray.js,
I found that the shouldComponentUpdate method is preventing the FieldArray to rerender

```
const propsToNotUpdateFor = [
  '_reduxForm',
  'value'
]
```

```
    shouldComponentUpdate(nextProps) {
      const nextPropsKeys = Object.keys(nextProps)
      const thisPropsKeys = Object.keys(this.props)
      return nextPropsKeys.length !== thisPropsKeys.length || nextPropsKeys.some(prop => {
        // useful to debug rerenders
        // if (!plain.deepEqual(this.props[ prop ], nextProps[ prop ])) {
        //   console.info(prop, 'changed', this.props[ prop ], '==>', nextProps[ prop ])
        // }
        return !~propsToNotUpdateFor.indexOf(prop) && !deepEqual(this.props[ prop ], nextProps[ prop ])
      })
    }
```

the value of the FieldArray's Field has changed, but it is returning false due to this part" !~propsToNotUpdateFor.indexOf(prop)"

**Comment by sckoh** [comment]
the value of the FieldArray's Field has changed, but it is returning false due to this part" !~propsToNotUpdateFor.indexOf(prop)"

**Comment by sckoh** [comment]
@erikras 
May I know what is the reason not rerender the FieldArray when value is changed?
I think that if the value is an object and it is changed, we should rerender the FieldArray right?

**Comment by dtipson** [comment]
I think I'm having a similar issue: deleting the item at index 0 of a FieldArray using ArrayRemove is resulting in the right input in the redux store, but wacky, mismatched display on the page (deleted item remains, items below it end up with mismatched props!)

**Comment by ashyap** [comment]
I'm also experiencing the same issue. Any workaround for this ?

**Comment by ashyap** [comment]
I found a workaround for this one. Just dispatch the arrayPush function instead of using fields.push

```
fields.remove(index)
dispatch(arrayPush('form', 'field', productToAdd))
```

**Comment by irvinlim** [comment]
I believe the `propsToNotUpdateFor` portion ([here](https://github.com/erikras/redux-form/blob/ba3b22e390137d11f344c739b3c7749a36d969c0/src/ConnectedFieldArray.js#L8)) as @sckoh mentioned was introduced in #1578 to prevent rerendering of the FieldArray when its Fields' values are being changed (see the issue for details).

However, this causes problems if the array `value`'s elements change and require render updates, which can result from `insert`, `remove`, `swap`, `move` etc. I personally am having trouble with getting `swap`s to appear correctly and consistently.

---

@erikras Could you offer some thoughts about this issue? I will be happy to create a PR if necessary to implement your suggestions if necessary.

My suggestion is to check `value` only for changes to the array length, or for changes to the order of the elements, in order to return `true` for `shouldComponentUpdate`.

Something like this:

```js
var thisValue = this.props.value;
var nextValue = nextProps.value;

if (thisValue && nextValue) {
  if (thisValue.length != nextValue.length || thisValue.every(val => nextValue.some(next => deepEqual(val, next)))) {
    return true
  }
}
```

**Comment by babsonmatt** [comment]
@dtipson that might be a key issue?

**Comment by dtipson** [comment]
Hmm, you're exactly right (Math.random keys fix the problem, sort of).  Though not sure what else to use as a key because nothing else is inherently unique and indeed, because of how FieldArray works, I don't even have access to the data to create a starting hash.

**Comment by erikras** [comment]
Fix published in [`v6.6.0`](https://github.com/erikras/redux-form/releases/tag/v6.6.0).

**Comment by stjepan** [comment]
...and so your code breaks now...

https://github.com/erikras/redux-form/issues/2740
